transports/onion/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+#![doc(html_logo_url = "https://libp2p.io/img/logo_small.png")]
+#![doc(html_favicon_url = "https://libp2p.io/img/favicon.png")]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+
+use std::pin::Pin;
+use std::sync::Arc;
+
+use address::{dangerous_extract_tor_address, safe_extract_tor_address};
+use arti_client::{TorAddrError, TorClient, TorClientBuilder};
+use futures::{future::BoxFuture, FutureExt};
+use libp2p_core::{transport::TransportError, Multiaddr, Transport};
+use tor_rtcompat::Runtime;
+
+mod address;
+mod provider;
+
+#[doc(inline)]
+pub use provider::OnionStream;
+
+#[derive(Debug, thiserror::Error)]
+pub enum OnionError {
+    #[error("error during address translation")]
+    AddrErr(#[from] TorAddrError),
+    #[error("error in arti")]
+    ArtiErr(#[from] arti_client::Error),
+}
+
+#[derive(Clone)]
+pub struct OnionClient<R: Runtime> {
+    // client is in an Arc, because wihtout it the Transport::Dial method can't be implemented,
+    // due to lifetime issues. With the, eventual, stabilization of static async traits this issue
+    // will be resolved.
+    client: Arc<TorClient<R>>,
+    pub conversion_mode: AddressConversion,
+}
+
+pub type OnionBuilder<R> = TorClientBuilder<R>;
+
+#[derive(Debug, Clone, Copy, Hash, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub enum AddressConversion {
+    /// uses only dns for address resolution
+    #[default]
+    DnsOnly,
+    /// uses ip and dns for addresses
+    IpAndDns,
+}
+
+impl<R: Runtime> OnionClient<R> {
+    pub fn from_builder(
+        builder: OnionBuilder<R>,
+        conversion_mode: AddressConversion,
+    ) -> Result<Self, OnionError> {
+        let client = Arc::new(builder.create_unbootstrapped()?);
+        Ok(Self {
+            client,
+            conversion_mode,
+        })
+    }
+
+    pub async fn bootstrap(&self) -> Result<(), OnionError> {
+        self.client.bootstrap().await.map_err(OnionError::ArtiErr)
+    }
+}
+
+macro_rules! default_constructor {
+    () => {
+        pub async fn default() -> Result<Self, OnionError> {
```

Comment:
Nit: I think people associate `default` with a very particular signature, the one from the `Default` trait. Maybe this is better as `new` or perhaps `bootstrapped`?

Reply:
`new` would fit very good in rust conventions, however `bootstrapped` would be better. I chose `default` exactly because it resembles the `Default` trait, which I would chose, if it would allow for async constructors.